File: bot.pl
#!/usr/bin/perl
use Digest::SHA qw/sha256_hex/;
use JSON;
use POSIX qw/floor/;
use List::Util qw/first/;
use POE;
use POE::Component::IRC;
use strict;

{ package Config; do 'config.pm' };

open CHAIN, 'chain.json';
my $chain = JSON->new->utf8->decode(<CHAIN>);
close CHAIN;

sub choice {
    my ($w, $n) = @_;
    my $l = @{$chain->{relations}->{$w}};
    my $i = floor($n * $l);
    return $chain->{relations}->{$w}->[$i];
}

sub indexOf {
    my ($v, $arr) = @_;

    my $index = first { $arr->[$_] eq $v } 0..$#{$arr};

    return $index;
}

sub calculate {
    my ($hashstr) = @_;

    # Start with a word after a period.
    my $w = indexOf('.', $chain->{wordlist});

    my $v = '';
    my $caps = 1;
    for (my $i = 0; $i < length($hashstr); $i += 2) {
        my $n = hex(substr($hashstr, $i, 2));
        my $wn = choice($w, $n / 256);
        my $word = $chain->{wordlist}->[$wn];
        if ($caps) {
            $word = ucfirst $word;
            $caps = 0;
        }
        if ($word =~ /[,.!?;:]/) {
            if ($word eq '.' || $word eq '!' || $word eq '?') {
                $caps = 1;
            }
        } else {
            $v .= ' ';
        }
        $v .= $word;
        $w = $wn;
    }

    $v =~ s/^ //;
    return $v . '.';
}

sub hash {
    my ($str) = @_;

    return calculate(sha256_hex($str));
}

my $irc = POE::Component::IRC->spawn;

POE::Session->create(
    inline_states => {
        _start => \&bot_start,
        irc_001 => \&on_connect,
        irc_public => \&on_public,
    }
);

sub bot_start {
    $irc->yield(register => 'all');
    $irc->yield(
        connect => {
            Nick => $Config::nick,
            Username => $Config::nick,
            Ircname => 'Down the Rabbit Hole',
            server => $Config::server,
            Port => $Config::port,
        }
    );
}

sub on_connect {
    for my $channel (@Config::channels) {
        $irc->yield(join => $channel);
    }
}

sub on_public {
    my ($kernel, $who, $where, $msg) = @_[KERNEL, ARG0, ARG1, ARG2];
    if ($msg =~ /^$Config::nick. (.+)/) {
        my $subject = $1;
        $irc->yield(privmsg => $where, hash($subject));
    }
}

$poe_kernel->run;
exit 0;

File: bot.py
#!/usr/bin/python
import hashlib
import json
import sys
import argparse
import time
from math import floor
from flow import Flow

import config

startTime = time.time()

f = open('chain.json')
chain = json.load(f)
f.close()

def choice(w, n):
    l = chain["relations"][str(w)]
    i = int(floor(n * len(l)))
    return chain["relations"][str(w)][i]

def calculate(hashstr):
    # Start with a word after a period.
    w = chain["wordlist"].index('.')

    v = ''
    caps = True
    for i in range(0, len(hashstr), 2):
        n = int(hashstr[i:i + 2], 16)
        wn = choice(w, n / 256.0)
        word = chain["wordlist"][wn]
        if caps:
            word = word[0].upper() + word[1:]
            caps = False
        if word in (',', '.', '!', '?', ';', ':'):
            if word in ('.', '!', '?'):
                caps = True
        else:
            v += ' '
        v += word
        w = wn

    return v.lstrip() + '.'

def hash(str):
    m = hashlib.new('sha256')
    m.update(str)
    return calculate(m.hexdigest())

backend = None
myAccountId = None

def start():
    global myAccountId, backend
    try:
        backend = Flow(config.username)
        print("Logged in as %s" % config.username)
    except Flow.FlowError as e:
        if str(e) == 'Unknown account':
            print("Account not found; attempting login")
            backend = Flow()
            try:
                backend.create_device(config.username, config.password)
                print("Logged in")
            except Flow.FlowError as e2:
                if str(e2) == 'blah':
                    print("Account appears to be non-existent; attempting to create")
                    backend.create_account(config.username, config.password)
                    backend.set_profile('profile', json.dumps({"displayName": config.displayname}))
                    print("Account created")
                else:
                    raise
        else:
            raise

    myAccountId = backend.account_id()

    @backend.message
    def process_message(notif_type, notif_data):
        regular_messages = notif_data["regularMessages"]
        for message in regular_messages:
            if message["creationTime"] / 1000 < startTime:
                continue

            try:
                otherdata = json.loads(message["otherData"])
            except ValueError:
                # Whoops, no otherdata, moving on...
                continue

            if myAccountId in otherdata["highlighted"]:
                sender = message["senderAccountId"]
                cid = message["channelId"]
                oid = backend.get_channel(cid)["orgId"]
                msg = message["text"]
                backend.send_message(oid, cid, hash(msg))

def main():
    start()

    parser = argparse.ArgumentParser(description='Markov Chain Bot')
    parser.add_argument('--join', type=str, help='A Team ID to join')

    args = parser.parse_args()

    if args.join:
        backend.new_org_join_request(args.join)
        print("Requested to join %s" % args.join)
        sys.exit(0)

    print("Ready")
    backend.process_notifications()

if __name__ == "__main__":
    main()
